File: M12 NF1 (Jose Java)/Softlearning/model/Products/Books.php
<?php

include($_SERVER['DOCUMENT_ROOT'].'/SoftLearning/model/Products/Product.php');
include($_SERVER['DOCUMENT_ROOT'].'/SoftLearning/model/Products/Dimensions.php');

include($_SERVER['DOCUMENT_ROOT'].'/SoftLearning/model/Products/Storable.php');
//include($_SERVER['DOCUMENT_ROOT'].'/SoftLearning/Exceptions/CheckData/BuildException.php');

class Books extends Product implements Storable {

    protected DateTime $date;
    protected string $author;
    protected string $isbn;
    protected string $cover;
    protected string $page;
    protected string $genre;
    protected string $editorial;
    protected Dimensions $Dim;
    protected bool $fragile;
    protected bool $create = true;

    public function __construct(string $ident, float $price, string $date,
            string $author, string $isbn, string $cover, string $page, string $genre, string $editorial,
            float $weight, float $height, float $width, float $length) {
        try {
            parent::__construct($ident, $price);
        } catch (BuildException $ex) {
            $errors .= $ex->getMessage();
        }

        try {
            $this->Dim = new Dimensions($weight, $height, $width, $length);
        } catch (Exception $ex) {
            throw new BuildException($ex->getMessage());
        }

        $errors = '';

        if ($this->setDate($date) !== 0) {
            $errors .= Check::getErrorMessage(Check::checkDate($date)) . "\n";
        }

        if ($this->setAuthor($author) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($author)) . "\n";
        }

        if ($this->setIsbn($isbn) !== 0) {
            $errors .= Check::getErrorMessage(Check::checkISBN($isbn)) . "\n";
        }

        if ($this->setCover($cover) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($cover)) . "\n";
        }

        if ($this->setPage($page) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($page)) . "\n";
        }

        if ($this->setGenre($genre) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($genre)) . "\n";
        }

        if ($this->setEditorial($editorial) !== 0) {
            $errors .= Check::getErrorMessage(Check::validate($editorial)) . "\n";
        }

        if (!empty($errors)) {
            throw new BuildException("Not possible to create the book(libro): \n" . $errors);
        }
    }
    
    
    public function getAuthor(): string {
        return $this->author;
    }
    public function getIsbn(): string {
        return $this->isbn;
    }

    public function getCover(): string {
        return $this->cover;
    }

    public function getPage(): string {
        return $this->page;
    }

    public function getGenre(): string {
        return $this->genre;
    }

    public function getEditorial(): string {
        return $this->editorial;
    }

    public function getDate(): string {
        return $this->date->format('Y-m-d');
    }

    public function getDim(): Dimensions {
        return $this->Dim;
    }

    public function setAuthor(string $author): int {
        $errorAuthor = Check::validate($author);
        if ($errorAuthor == 0) {
            $this->author = $author;
        }
        return $errorAuthor;
    }

    public function setDate(string $date): int {
        $errorDate = Check::checkDate($date);
        if ($errorDate == 0) {
            $this->date = new DateTime($date);
        }
        return $errorDate;
    }

    public function setIsbn(string $isbn): int {
        $errorIsbn = Check::checkISBN($isbn);
        if ($errorIsbn == 0) {
            $this->isbn = $isbn;
        }
        return $errorIsbn;
    }

    public function setCover(string $cover): int {
        $errorCover = Check::validate($cover);
        if ($errorCover == 0) {
            $this->cover = $cover;
        }
        return $errorCover;
    }

    public function setPage(string $page): int {
        $errorPage = Check::validate($page);
        if ($errorPage == 0) {
            $this->page = $page;
        }
        return $errorPage;
    }

    public function setGenre(string $genre): int {
        $errorGenre = Check::validate($genre);
        if ($errorGenre == 0) {
            $this->genre = $genre;
        }
        return $errorGenre;
    }

    public function setEditorial(string $editorial): int {
        $errorEditorial = Check::validate($editorial);
        if ($errorEditorial == 0) {
            $this->editorial = $editorial;
        }
        return $errorEditorial;
    }

    public function setDim(Dimensions $Dim): void {
        $this->Dim = $Dim;
    }

    public function isFragile(): bool {
        return $this->fragile;
    }

    public function volume(): float {
        return $this->Dim->getVolume();
    }

    public function getDetails(): string {
        return "ID: " . $this->getIdent() . "\n" .
                "Price: $" . number_format($this->getPrice(), 2) . "\n" .
                "Author: " . $this->getAuthor() . "\n" .
                "ISBN: " . $this->getIsbn() . "\n" .
                "Cover: " . $this->getCover() . "\n" .
                "Pages: " . $this->getPage() . "\n" .
                "Genre: " . $this->getGenre() . "\n" .
                "Editorial: " . $this->getEditorial() . "\n" .
                "Date: " . $this->getDate() . "\n" .
                "Dimensions (WxHxL): " . $this->getDim()->getWidth() . "x" . $this->getDim()->getHeight() . "x" . $this->getDim()->getLength() . " cm\n" .
                "Weight: " . $this->getDim()->getWeight() . " kg\n";
    }

    public function isFlexible(): bool {
        return $this->isFlexible();
    }

    public function isHeavy(): bool {
        return $this->isHeavy();
    }
}

File: M12 Jose Java/softlearningmaven/src/main/java/com/core/entities/book/model/Books.java
package com.core.entities.book.model;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

import com.core.checks.Check;
import com.core.entities.shared.dimensions.Dimensions;
import com.core.entities.shared.products.Product;
import com.core.entities.shared.storable.Storable;


public class Books extends Product implements Storable {
    
    protected LocalDate date;
    protected DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
    protected String author;
    protected String isbn;
    protected String cover;
    protected int page;
    protected String genre;
    protected String editorial;
    protected Dimensions dim;
    protected boolean fragile;
    
    protected Books(){
    
    }
    
    public static Books getInstance(
        String ident, 
        double price,
        boolean delayPay, 
        double discount, 
        String type, 
        String payMethod, 
        String date, 
        String author, 
        String isbn, 
        String cover, 
        int page, 
        String genre, 
        String editorial, 
        double weight, 
        double height, 
        double width, 
        boolean fragile, 
        double length) throws Exception {
        
        StringBuilder errors = new StringBuilder();
        int errorCode;
        Books libro1 = new Books();
        try {
            libro1.dim = Dimensions.getInstanceDimensions(weight, height, width, fragile, length); 
        } catch (Exception e) {
            throw new Exception("Error en las dimensiones: " + e.getMessage()); 
        }
        
        try {
            libro1.product(ident, price, delayPay, discount, type, payMethod);
            
        } catch (Exception ex) {
            throw new Exception("Error en product: " + ex.getMessage());
        }

        
        if ((errorCode = libro1.setDate(date)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }
        if ((errorCode = libro1.setAuthor(author)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }
        if ((errorCode = libro1.setIsbn(isbn)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }
        if ((errorCode = libro1.setCover(cover)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }
        if ((errorCode = libro1.setPage(page)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }
        if ((errorCode = libro1.setGenre(genre)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }
        if ((errorCode = libro1.setEditorial(editorial)) != 0) {
            errors.append(Check.getErrorMessage(errorCode)).append("\n");
        }

        
        if (errors.length() > 0) {
            throw new Exception("No es posible crear el libro: \n" + errors.toString());

        }
        return libro1;
    }
    

    


    // getter
    public String getAuthor() {
        return this.author;
    }

    public String getIsbn() {
        return this.isbn;
    }

    public String getCover() {
        return this.cover;
    }

    public int getPage() {
        return this.page;
    }

    public String getGenre() {
        return this.genre;
    }

    public String getDate() {
        return this.date.format(formatter);
    }

    public String getEditorial() {
        return this.editorial;
    }

    public Dimensions getDim() {
        return this.dim;
    }

    public double getWeight(){
        return dim.getWeight();
    }
    public double getHeight() {
        return dim.getHeight();
    }
    
    public double getWidth() {
        return dim.getWidth();
    }
    
    public boolean getFragile() {
        return this.fragile;
    }

    public double getLength() {
        return dim.getLength();
    }
    public double getVolume() {
        return dim.getLength() * dim.getWidth() * dim.getHeight();
    }


    

    // setter
    public int setAuthor(String author) {
        int errorAuthor = Check.minMaxLength(author);
        if (errorAuthor == 0) {
            this.author = author;
        }
        return errorAuthor;
    }

    public int setDate(String date) {
        int errorDate = Check.isValidDate(date);  // Usamos isValidDate para validar
        
        if (errorDate == 0) {
            // Si es valida, la analizamos y asignamos
            this.date = LocalDate.parse(date, this.formatter);
        }
        
        return errorDate;
    }



//     public int setDate(String date) {
//         int errorDate = Check.isValidDate(date);
//         if (errorDate == 0) {
//             this.date = LocalDate.parse(date, formatter);
//         }
//         return errorDate;
// }
    

    public int setIsbn(String isbn) {
        int errorIsbn = Check.checkISBN(isbn);
        if (errorIsbn == 0) {
            this.isbn = isbn;
        }
        return errorIsbn;
    }

    public int setCover(String cover) {
        int errorCover = Check.minMaxLength(cover);
        if (errorCover == 0) {
            this.cover = cover;
        }
        return errorCover;
    }

    public int setPage(int page) {
        int errorPage = Check.range(page);
        if (errorPage == 0) {
            this.page = page;
        }
        return errorPage;
    }

    public int setGenre(String genre) {
        int errorGenre = Check.minLenght(genre);
        if (errorGenre == 0) {
            this.genre = genre;
        }
        return errorGenre;
    }

    public int setEditorial(String editorial) {
        int errorEditorial = Check.minLenght(editorial);
        if (errorEditorial == 0) {
            this.editorial = editorial;
        }
        return errorEditorial;
    }

    

    public void setDim(Dimensions dim) {
        this.dim = dim;
    }
    
    

    
    @Override
    public double volume() {
        return this.dim.getVolume();
    }

    @Override
    public String getDetails() {
        return "ID: " + this.getIdent() + "\n" +
                "Precio: $" + this.getPrice() + "\n" +
                "Autor: " + this.getAuthor() + "\n" +
                "ISBN: " + this.getIsbn() + "\n" +
                "Tapa: " + this.getCover() + "\n" +
                "Paginas: " + this.getPage() + "\n" +
                "Género: " + this.getGenre() + "\n" +
                "Editorial: " + this.getEditorial() + "\n" +
                "Fecha: " + this.getDate() + "\n" +
                "Dimensiones (AxAxL): " + this.getDim().getWidth() + "x" + this.getDim().getHeight() + "x" + this.getDim().getLength() + " cm\n" +
                "Peso: " + this.getDim().getWeight() + " kg\n";
    }

    @Override
    public boolean isFlexible() {
        return false; 
    }

    @Override
    public boolean isHeavy() {
        return this.dim.getWeight() > 1.0 ? true:false; //si pesa mas de 1kg
    }

    
}
